File: internal/journal/store.go
package journal

import "time"

// Store persists entries, keyed by calendar date.
//
// Implementations must be safe for use from a single goroutine; the app drives
// all access from the UI goroutine. Every method returns an error rather than
// panicking, because a failure on a user's phone must become a dialog, not a
// crash.
type Store interface {
	// Get returns the entry for d. The bool reports whether one exists; a
	// missing entry is not an error.
	Get(d Date) (Entry, bool, error)

	// Put writes e, replacing any existing entry for e.Date.
	Put(e Entry) error

	// PutAll writes every entry atomically: either all of them land or, if the
	// write fails, none do. It is what makes CSV import all-or-nothing.
	PutAll(entries []Entry) error

	// Delete removes the entry for d. Deleting a date with no entry is not an
	// error.
	Delete(d Date) error

	// OnThisDay returns every entry falling on the given month and day, in any
	// year present, ordered newest year first. It includes the current year if
	// an entry exists; filtering that out belongs to the caller.
	OnThisDay(month time.Month, day int) ([]Entry, error)

	// Dates returns every date that has an entry, in ascending order.
	Dates() ([]Date, error)

	// All calls yield for every entry in ascending date order. If yield returns
	// an error, iteration stops and All returns that error.
	All(yield func(Entry) error) error

	// Close releases the underlying resources.
	Close() error
}

File: src/domain/store.ts
import type { JournalDate } from './date';
import type { Entry } from './entry';

/**
 * Store persists entries, keyed by calendar date.
 *
 * Every method rejects with an Error rather than throwing synchronously or
 * crashing, because a failure on a user's phone must become an alert, not a
 * dead app.
 *
 * Implementations are driven from the UI thread and need not be safe against
 * concurrent callers — but note that "single-threaded" does not mean
 * "uninterleaved" in JavaScript the way it does for the Go original's single
 * goroutine: two overlapping awaits on the same store genuinely interleave.
 * Implementations must therefore keep each method's own writes atomic
 * (see putAll) rather than assuming no other call is in flight.
 */
export interface Store {
  /** get returns the entry for date, or null. A missing entry is not an error. */
  get(date: JournalDate): Promise<Entry | null>;

  /** put writes entry, replacing any existing entry for the same date. */
  put(entry: Entry): Promise<void>;

  /**
   * putAll writes every entry atomically: either all of them land or, if the
   * write fails, none do. It is what makes CSV import all-or-nothing.
   * An empty array is a no-op, not an error.
   */
  putAll(entries: Entry[]): Promise<void>;

  /**
   * delete removes the entry for date. Deleting a date that has no entry is
   * not an error.
   */
  delete(date: JournalDate): Promise<void>;

  /**
   * onThisDay returns every entry falling on the given 1-based month and day,
   * in any year present, ordered newest year first. It includes the current
   * year if an entry exists; filtering that out is the Memories screen's job.
   */
  onThisDay(month: number, day: number): Promise<Entry[]>;

  /** dates returns every date that has an entry, in ascending order. */
  dates(): Promise<JournalDate[]>;

  /**
   * all yields every entry in ascending date order. A consumer that stops
   * iterating (break, throw, or an early return) stops the underlying read.
   */
  all(): AsyncIterable<Entry>;

  /** close releases the underlying resources. */
  close(): Promise<void>;
}
